examples/end2end_tree_reconstruction_test.py

```diff
@@ -0,0 +1,243 @@
+    fig.set_size_inches(20, 20)
+    plt.tight_layout()
+
+
+def visualize_reconstruction(
+    true_phylo_df: pd.DataFrame,
+    reconst_phylo_df: pd.DataFrame,
+    *,
+    visualize: bool,
+    **kwargs,
+) -> None:
+    """Print a sample of the reference and reconstructed phylogenies."""
+    show_taxa = (
+        reconst_phylo_df["taxon_label"].dropna().sample(6, random_state=1)
+    )
+    print("ground-truth phylogeny sample:")
+    print(to_ascii(true_phylo_df, show_taxa))
+    print()
+    print("reconstructed phylogeny sample:")
+    print(to_ascii(reconst_phylo_df, show_taxa))
+
+    if visualize:
+        true_phylo_df["taxon_label"] = true_phylo_df["taxon_label"].apply(
+            lambda x: x[:5]
+        )
+        reconst_phylo_df["taxon_label"] = reconst_phylo_df[
+            "taxon_label"
+        ].apply(lambda x: x and x[:5])
+        tp.tee(
+            generate_phylogeny_images,
+            alifestd_try_add_ancestor_list_col(true_phylo_df),
+            alifestd_try_add_ancestor_list_col(reconst_phylo_df),
+            teeplot_outattrs=kwargs,
+            teeplot_outdir="/tmp",
+        )
+
+
+def test_reconstruct_one(
+    differentia_bitwidth: int,
+    surface_size: int,
+    fossil_interval: typing.Optional[int],
+    *,
+    visualize: bool,
+) -> float:
+    """Test the reconstruction of a single phylogeny."""
+    print("=" * 80)
+    print(f"surface_size: {surface_size}")
+    print(f"differentia_bitwidth: {differentia_bitwidth}")
+    print(f"fossil_interval: {fossil_interval}")
+
+    true_phylo_df, reconst_phylo_df = sample_reference_and_reconstruction(
+        differentia_bitwidth,
+        surface_size,
+        fossil_interval,
+    )
+
+    visualize_reconstruction(
+        true_phylo_df,
+        reconst_phylo_df,
+        differentia_bitwidth=differentia_bitwidth,
+        surface_size=surface_size,
+        fossil_interval=fossil_interval,
+        visualize=visualize,
+    )
+    reconstruction_error = alifestd_calc_triplet_distance_asexual(
+        alifestd_collapse_unifurcations(true_phylo_df), reconst_phylo_df
+    )
+    print(f"{reconstruction_error=}")
+    assert reconstruction_error < alifestd_count_leaf_nodes(true_phylo_df)
+
+    return reconstruction_error
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--visualize", action="store_true")
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    sys.setrecursionlimit(100000)
+    args = _parse_args()
+    reconstruction_errors = [
+        test_reconstruct_one(
+            differentia_bitwidth,
+            surface_size,
+            fossil_interval,
+            visualize=args.visualize,
+        )
+        for fossil_interval in (None, 50, 200)
+        for surface_size in (256, 64, 16)
+        for differentia_bitwidth in (64, 8, 1)
+    ]
+    # error should increase with decreasing surface size
+    assert all(
+        sorted(reconstruction_errors[i : i + 9])
```

Comment:
more itertools (mit) chunked; if we collate results as a data frame, this could be done by iterating through a groupby